nodejs/lib/tensor-impl.ts

```diff
@@ -0,0 +1,170 @@
+        if (constructor === undefined) {
+          throw new TypeError(`Unknown tensor type: ${arg0}.`);
+        }
+        if (Array.isArray(arg1)) {
+          // use 'as any' here because TypeScript's check on type of 'SupportedTypedArrayConstructors.from()' produces
+          // incorrect results.
+          // 'constructor' should be one of the typed array prototype objects.
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          data = (constructor as any).from(arg1);
+        } else if (arg1 instanceof constructor) {
+          data = arg1;
+        } else {
+          throw new TypeError(`A ${type} tensor's data must be type of ${constructor}`);
+        }
+      }
+    } else {
+      //
+      // Override: constructor(data, ...)
+      //
+      dims = arg1;
+      if (Array.isArray(arg0)) {
+        // only boolean[] and string[] is supported
+        if (arg0.length === 0) {
+          throw new TypeError('Tensor type cannot be inferred from an empty array.');
+        }
+        const firstElementType = typeof arg0[0];
+        if (firstElementType === 'string') {
+          type = 'string';
+          data = arg0;
+        } else if (firstElementType === 'boolean') {
+          type = 'bool';
+          // 'arg0' is of type 'boolean[]'. Uint8Array.from(boolean[]) actually works, but typescript thinks this is
+          // wrong type. We use 'as any' to make it happy.
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          data = Uint8Array.from(arg0 as any[]);
+        } else {
+          throw new TypeError(`Invalid element type of data array: ${firstElementType}.`);
+        }
+      } else {
+        // get tensor type from TypedArray
+        const mappedType = NUMERIC_TENSOR_TYPEDARRAY_TO_TYPE_MAP.get(arg0.constructor);
+        if (mappedType === undefined) {
+          throw new TypeError(`Unsupported type for tensor data: ${arg0.constructor}.`);
+        }
+        type = mappedType;
+        data = arg0 as SupportedTypedArray;
+      }
+    }
+
+    // type and data is processed, now processing dims
+    if (dims === undefined) {
+      // assume 1-D tensor if dims omitted
+      dims = [data.length];
+    } else if (!Array.isArray(dims)) {
+      throw new TypeError('A tensor\'s dims must be a number array');
+    }
+
+    // perform check
+    const size = calculateSize(dims);
+    if (size !== data.length) {
+      throw new Error(`Tensor's size(${size}) does not match data length(${data.length}).`);
+    }
+
+    this.dims = dims as readonly number[];
+    this.type = type;
+    this.data = data;
+    this.size = size;
+  }
+  //#endregion
+
+  //#region fields
+  readonly dims: readonly number[];
+  readonly type: TensorInterface.Type;
+  readonly data: TensorInterface.DataType;
+  readonly size: number;
+  //#endregion
+
+  //#region tensor utilities
+  reshape(dims: readonly number[]): TensorInterface {
+    return new Tensor(this.type, this.data, dims);
```

Comment:
Why do we need to support creating a new "reshaped" tensor with the same underlying data ? ORT native Tensor class has Reshape() that just reshapes the dims of the current Tensor (after validating compatiblity).

Reply:
I made all this assumption - a `Tensor` object, once created, its type and dims should be constant.

ORT native tensor and its functions are implementation details and are actualy not exposed to users. (what users see is a more abstract type OrtValue).

creating a Tensor object in javascript in relatively cheep, and what happened inside is a shallow copy of buffer and copy of metadata (type/dims).

the 'utilities' region, as planned, will contain a set of utility functions to help users to use onnxruntime easier. so far there is only one function; but there will be more in future. if any of those function need to modify the metadata of a tensor, it will always create a new tensor object.